alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/implementations/conditions/BiomolPresentInEnv.java

```diff
@@ -8,28 +8,106 @@
  *
  */
 public class BiomolPresentInEnv extends GenericMoleculePresent<Double> {
 
-    private static final long serialVersionUID = 6527297022655890961L;
+    private static final long serialVersionUID = 1L;
+
+    private final Environment<Double> environment;
 
     /**
+     * Initialize condition for extra-cellular environment, implemented as a set
+     * of {@link EnvironmentNode}.
      * 
-     * @param biomol 
-     * @param concentration 
-     * @param node 
+     * @param biomol
+     *            the {@link Biomolecule} which the condition is about.
+     * @param conc
+     *            the requested concentration.
+     * @param node
+     *            the node where this condition is located;
+     * @param env
+     *            the {@link Environment} where the node is located.
      */
-    public BiomolPresentInEnv(final Biomolecule biomol, final Double concentration, final EnvironmentNode node) {
-        super(biomol, node, concentration);
+    public BiomolPresentInEnv(final Environment<Double> env, final Node<Double> node, final Biomolecule biomol, final Double conc) {
+        super(node, biomol, conc);
+        environment = env;
     }
 
     @Override
-    public EnvironmentNode getNode() {
-        return (EnvironmentNode) super.getNode();
+    public double getPropensityConditioning() {
+        double quantityInEnvNodes = 0;
+        if (!getEnviromentNodesSurrounding().isEmpty()) {
+            quantityInEnvNodes = getEnviromentNodesSurrounding().stream()
+                    .parallel()
+                    .mapToDouble(n -> n.getConcentration(getBiomolecule()))
+                    .sum();
+        }
+        double quantityInLayers = 0;
+        if (environment.getLayer(getBiomolecule()).isPresent()) {
+            quantityInLayers = environment.getLayer(getBiomolecule()).get().getValue(environment.getPosition(getNode()));
+        }
+        final double totalQuantity = quantityInEnvNodes + quantityInLayers;
+        if (totalQuantity < getQuantity()) {
+            return 0;
+        }
+        return CombinatoricsUtils.binomialCoefficientDouble(
+                (int) FastMath.round(totalQuantity), 
+                (int) FastMath.round(getQuantity())
+                );
+    }
+
+    /**
+     * @return a list of EnvironmentNodes near to the node where this condition is located.
+     */
+    protected final List<Node<Double>> getEnviromentNodesSurrounding() {
+        return environment.getNeighborhood(getNode()).getNeighbors().stream()
+                .parallel()
+                .filter(n -> n instanceof EnvironmentNode)
+                .collect(Collectors.toList());
+    }
+
+    @Override 
+    public BiomolPresentInEnv cloneOnNewNode(final Node<Double> n) {
+        return new BiomolPresentInEnv(environment, n, getBiomolecule(), getQuantity());
+    }
+
+    @Override
+    public Context getContext() {
+        return Context.NEIGHBORHOOD;
+    }
+
+    @Override
+    public boolean isValid() {
```

Comment:
This method and `getPropensityConditioning` are copy/pasted. Refactor (e.g. write a private method that hosts the common code parts).